src/java/org/apache/cassandra/index/sai/analyzer/ByteLimitedMaterializer.java

```diff
@@ -0,0 +1,93 @@
+
+import static org.apache.cassandra.config.CassandraRelevantProperties.SAI_MAX_ANALYZED_TERM_SIZE;
+
+/**
+ * Utility class for conditionally materializing a list of tokens given an analyzer and the term to be analyzed.
+ * If the cumulative size of the analyzed term exceeds the configured maximum, an empty list is returned to prevent
+ * excessive memory/disk usage for a single term.
+ */
+public class ByteLimitedMaterializer
+{
+    private static final Logger logger = LoggerFactory.getLogger(ByteLimitedMaterializer.class);
+    private static final NoSpamLogger noSpamLogger = NoSpamLogger.getLogger(logger, 1, TimeUnit.MINUTES);
+    public static final int MAX_ANALYZED_TERM_SIZE = SAI_MAX_ANALYZED_TERM_SIZE.getInt();
+    public static final String ANALYZED_TERM_OVERSIZE_MESSAGE = "Cannot add term's analyzed tokens of column {} to index" +
+                                                                " for key: {}, analzyed term size {}, but max allowed size {}.";
+
+    /**
+     * Using the configured analyzer, materialize the tokens for the given term. If the cumulative size of the analyzed
+     * term exceeds the configured maximum, an empty list is returned to prevent excessive memory/disk usage for a
+     * single term. If the analyzer does not transform the value, the analyzer's byte limit is ignored because indexes
+     * already have a limit on the size of a single term.
+     * @param analyzer the analyzer to use
+     * @param term the term to analyze
+     * @param indexContext the index context used for logging
+     * @param primaryKey the primary key of the row being indexed
+     * @return all the terms produced by the analyzer, or an empty list if the cumulative size of the analyzed term
+     * exceeds the configured maximum
+     */
+    public static List<ByteBuffer> materializeTokens(AbstractAnalyzer analyzer, ByteBuffer term, IndexContext indexContext, PrimaryKey primaryKey)
+    {
+        try
+        {
+            analyzer.reset(term);
+            if (!analyzer.transformValue())
+                return analyzer.hasNext() ? List.of(analyzer.next()) : List.of();
+
+            List<ByteBuffer> tokens = new ArrayList<>();
+            int bytesCount = 0;
+            while (analyzer.hasNext())
+            {
+                final ByteBuffer token = analyzer.next();
+                tokens.add(token);
+                bytesCount += token.remaining();
+                if (bytesCount >= MAX_ANALYZED_TERM_SIZE)
+                {
+                    noSpamLogger.warn(indexContext.logMessage(ANALYZED_TERM_OVERSIZE_MESSAGE),
+                                      indexContext.getColumnName(),
+                                      indexContext.keyValidator().getString(primaryKey.partitionKey().getKey()),
+                                      FBUtilities.prettyPrintMemory(bytesCount),
+                                      FBUtilities.prettyPrintMemory(MAX_ANALYZED_TERM_SIZE));
+                    return List.of();
```

Comment:
Can we take the approach we do elsewhere, where if we're inserting into an indexed column we will throw invalid request, and if we're indexing existing data we'll log and skip?